vllm/logits_process.py

```diff
@@ -0,0 +1,119 @@
+from typing import Callable, List, Tuple, Union
+
+import torch
+
+from vllm.transformers_utils.tokenizer import AnyTokenizer, MistralTokenizer
+
+LogitsProcessor = Union[Callable[[List[int], torch.Tensor], torch.Tensor],
+                        Callable[[List[int], List[int], torch.Tensor],
+                                 torch.Tensor]]
+"""LogitsProcessor is a function that takes a list
+of previously generated tokens, the logits tensor
+for the next token and, optionally, prompt tokens as a
+first argument, and returns a modified tensor of logits
+to sample from."""
+
+
+def get_bad_words_logits_processors(
+        bad_words: List[str],
+        tokenizer: AnyTokenizer) -> List[LogitsProcessor]:
+    bad_words_ids: List[List[int]] = list()
+
+    for bad_word in bad_words:
+        # To prohibit words both at the beginning
+        # and in the middle of text
+        # (related to add_prefix_space tokenizer parameter)
+        for add_prefix_space in [False, True]:
+            prefix = " " if add_prefix_space else ""
+            inputs = {"prompt": prefix + bad_word.lstrip()}
+
+            if isinstance(tokenizer, MistralTokenizer):
+                # Mistral tokenizers should not add special tokens
+                prompt_token_ids = tokenizer.encode(prompt=inputs["prompt"])
+            else:
+                prompt_token_ids = tokenizer.encode(text=inputs["prompt"],
+                                                    add_special_tokens=False)
+
+            # If no space at the beginning
+            # or if prefix space produces a new word token
+            if (not add_prefix_space) or (
+                    add_prefix_space
+                    and prompt_token_ids[0] != bad_words_ids[-1][0]
+                    and len(prompt_token_ids) == len(bad_words_ids[-1])):
+                bad_words_ids.append(prompt_token_ids)
+
+    return [NoBadWordsLogitsProcessor(bad_words_ids=bad_words_ids)]
+
+
+class NoBadWordsLogitsProcessor:
+    _SMALLEST_LOGIT = float("-inf")
+    _NEUTRAL_LOGIT = 0.0
+
+    def __init__(self, bad_words_ids: List[List[int]]):
+        self.bad_words_ids = bad_words_ids
+        self.word_bias: torch.FloatTensor = None
+
+    def __call__(
+        self,
+        past_tokens_ids: Union[List[int], Tuple[int]],
+        logits: torch.FloatTensor,
+    ) -> torch.Tensor:
+        if self.word_bias is None:
+            self._init_word_bias(logits=logits)
+
+        last_token_bias = torch.zeros_like(logits)
+
+        for bad_word_ids in self.bad_words_ids:
+            if len(bad_word_ids) == 1:  # 1-token words already processed
+                continue
+
+            if len(bad_word_ids) > len(past_tokens_ids) + 1:
```

Comment:
This differs from the original inequality here: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L1131
```python
if len(sequence_ids) > input_ids.shape[1]:  # the sequence is longer than the context, ignore
    continue
```

I may be wrong, but it seemed we should allow for one token to be generated (so + 1).

Reply:
According to your interpretation, we should also change `prefix_length` to `len(bad_word_ids)`. See if your tests still pass with this change.

Reply:
One passes, one fails:

<img src="https://github.com/user-attachments/assets/e61d3613-5c2f-4dd9-bfbb-d7d4b7cf9150" width="80%">

Well, I think, I better try come up with some illustration of the idea (to make the picture clearer for everybody, including me :slightly_smiling_face:)

Suppose, `past_tokens_ids = ["bye"]` (well, that's not ids, but words — let's go with it for simplicity).
And `bad_word_ids = ["bye", "world"]`.
So, we do not want "bye world" to be generated, but we already have "bye" generated.

What is happening now:
```python
past_tokens_ids = ["bye"]
bad_word_ids = ["bye", "world"]

if len(bad_word_ids) > len(past_tokens_ids) + 1:  # 2 > 1 + 1 -> False
                                                  # so we are not skipping `bad_word_ids`
    continue

prefix_length = len(bad_word_ids) - 1             # 2 - 1 = 1
last_token_id = bad_word_ids[-1]                  # "world"
actual_prefix = past_tokens_ids[-prefix_length:]  # ["bye"]
expected_prefix = bad_word_ids[:prefix_length]    # ["bye"]

assert len(actual_prefix) == len(expected_prefix)  # 1 == 1

is_match = tuple(actual_prefix) == tuple(expected_prefix)  #  ["bye"] ==  ["bye"]
last_token_bias[last_token_id] += (
    self._SMALLEST_LOGIT if is_match
    else self._NEUTRAL_LOGIT)  # "world" gets float("-inf") as logit
```

So, the word "world" is prohibited.

Reply:
I see, thanks for the detailed explanation.